### ISSUE TITLE
Training refactoring

### ISSUE THREAD

**Issue by constantinpape** [computational-cell-analytics/micro-sam#141]
We should refactor the training logic, so that the `SamTrainer` takes a prompt generator, which then takes care of all the iterative prompting logic (or not for a simpler training scheme, in which case also `n_sub_iterations=0`).

Now that we have the `IterativePromptingGenerator` that should be feasible. In order to make this extensible we need to define clean signatures for the `__call__` method of the prompt generators, and for their return values.

And we should then also clean up and simplify the training and iterative prompt generator code. (There's still some inefficient things in there.)

**Comment by anwai98** [comment]
Hi Constantin. The PR - #192, addresses the aforementioned issues with some basic cleaning of the trainer logic. The next steps would be taking this forward step-wise to take care of further concerns (as already discussed). Thanks.

**Comment by constantinpape** [comment]
I have merged #192 now @anwai98. The next concrete steps are:
- Implementing tests for the iterative prompt generator and the training function. (So that we have some checks that future PRs that work on this don't break this functionality).
- Further refactor the prompt generators, so that we can define a base class and common interface, s.t. the generator design becomes modular and can be used in the training as well as inference function.
    - Based on this design, adapt the training and inference that "batch" and iterative prompting can be used.
-  Refactor more training functionality to work on batches instead of using loops.
- Adapt more training scripts to this, add them here, train more generalists.

Not necessarily in that order ;). We can discuss later how to tackle this and how to split the work.

**Comment by constantinpape** [comment]
Another thing I noticed: we set `channelwise=False` in the dice loss, e.g.
https://github.com/computational-cell-analytics/micro-sam/blob/master/finetuning/livecell_finetuning.py#L64-L65

I am not sure this is a good idea, we should revisit this while working on the training.

**Comment by constantinpape** [comment]
I started with the tests in #198

**Comment by anwai98** [comment]
With the most recent changes, looks like the performance quality can be reproduced (on LiveCell) with the updates in the training scheme.

**Comment by constantinpape** [comment]
This has been implemented thanks to the work of @anwai98
